src/wp-includes/block-editor.php

```diff
@@ -243,6 +243,37 @@ function get_block_editor_settings( $editor_name, $custom_settings = array() ) {
 		$custom_settings
 	);
 
+	$editor_settings['__experimentalFeatures'] = WP_Theme_JSON_Resolver::get_merged_data( $editor_settings )->get_settings();
```

Comment:
By passing the "global settings" through this key, we have a natural mechanism to filter them from the plugin: using the existing block editor filters. If/when the plugin introduces some changes to the settings retrieval, this can work this way:

- core pulls data from core & the theme and generates `__experimentalFeatures`
- the plugin hooks into `block_editor_settings(_all)` to do the modifications it needs

Note that we need to cover against one use case: that the theme in use has the settings provided in a format different from what core understands. In that case, core ignores the data ([see](https://github.com/WordPress/wordpress-develop/pull/1262/files#diff-e12c3008d747d1bec5be9927c5e7b1ced0a88641abe52c278d495da936714817R68)) and it's the plugin responsibility to fill it up.

Reply:
Complementary to this, I've also added a specific filter `theme_json_allowed_settings` so consumers can add/remove the settings for WordPress core to process normally.

Reply:
Is this call done before or after the `block_editor_settings` filter? Because If I remember properly `get_merged_data` use the settings, so plugins should be able to change the old settings before being passed to `get_merged_data`?

Reply:
I actually have [written down the conclusions of that conversation](https://github.com/WordPress/gutenberg/issues/30082#issuecomment-816679803) but then forgot about it. It seems that the bit we're missing is this:

> Use the top-level settings of theme.json to set the current block editor settings. The settings per block are still stored within __experimentalFeatures.

Parsing that comment I think this PR needs to:

- unset the top-level settings within `__experimentalFeatures` that match an existing setting. For example, unset `settings.__experimentalFeatures.color.palette` because that value is already `settings.colors` and 3rd-parties can use filter it. The `useSetting` hook will fall back to the proper setting.
- the new top-level settings remain in `__experimentalFeatures` as well as the block settings.

Is this correct?

cc @jorgefilipecosta you were asking some questions about this.

Reply:
- unset the top-level settings within __experimentalFeatures that match an existing setting. For example, unset settings.__experimentalFeatures.color.palette because that value is already settings.colors and 3rd-parties can use filter it. The useSetting hook will fall back to the proper setting.
- the new top-level settings remain in __experimentalFeatures as well as the block settings.


This approach causes some friction a plugin developer that wants to change some settings will need to know what settings are old and in that case will change the old settings.colors and which ones are new and in that case use __experimentalFeatures.color.palette.
It seems strange to have two disjoint systems to deal with the same thing.

But there is no easy solution a different approach I see is to call apply filters on settings passed to WP_Theme_JSON_Resolver::get_merged_data( $editor_settings ). In that case, filters would be applied two times which is also not ideal.

Reply:
Done https://github.com/WordPress/wordpress-develop/pull/1262/commits/d4a16a3696c9c80d87d336b961378010edb74fd7

The way I see it is: consumers (and Gutenberg ) still use the existing settings and we introduce _more_ settings that are experimental.